File: iOS/Sources/VertuEdge/HuggingFace/ModelDownloadViewModel.swift
import Foundation

@MainActor
class ModelDownloadViewModel: ObservableObject {
    @Published var models: [ModelInfo] = []
    @Published var downloadState: DownloadState = .idle

    private let repository = HuggingFaceRepository()

    func searchModels(query: String) {
        Task {
            do {
                let results = try await repository.searchModels(query: query)
                models = results + (await repository.getLocalModels())
            } catch {
                models = await repository.getLocalModels()
            }
        }
    }

    func downloadModel(_ model: ModelInfo) {
        downloadState = .downloading(modelName: model.name, progress: 0)
        Task {
            do {
                let path = try await repository.downloadModel(model: model) { progress in
                    Task { @MainActor in
                        self.downloadState = .downloading(modelName: model.name, progress: progress)
                    }
                }
                downloadState = .completed(modelName: model.name, localPath: path)
            } catch {
                downloadState = .error(message: error.localizedDescription)
            }
        }
    }
}

File: Android/app/src/main/java/com/vertu/edge/huggingface/ModelDownloadViewModel.kt
package com.vertu.edge.huggingface

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class ModelDownloadViewModel @Inject constructor(
    private val repository: HuggingFaceRepository
) : ViewModel() {

    private val _models = MutableStateFlow<List<ModelInfo>>(emptyList())
    val models: StateFlow<List<ModelInfo>> = _models

    private val _downloadState = MutableStateFlow<DownloadState>(DownloadState.Idle)
    val downloadState: StateFlow<DownloadState> = _downloadState

    init {
        searchModels("")
    }

    fun searchModels(query: String) {
        viewModelScope.launch {
            repository.searchModels(query).onSuccess { results ->
                _models.value = results + repository.getLocalModels()
            }.onFailure {
                _models.value = repository.getLocalModels()
            }
        }
    }

    fun downloadModel(model: ModelInfo) {
        viewModelScope.launch {
            _downloadState.value = DownloadState.Downloading(model.name, 0f)
            repository.downloadModel(model) { progress ->
                _downloadState.value = DownloadState.Downloading(model.name, progress)
            }.onSuccess { path ->
                _downloadState.value = DownloadState.Completed(model.name, path)
            }.onFailure { e ->
                _downloadState.value = DownloadState.Error(e.message ?: "Download failed")
            }
        }
    }
}
